File: scripts/process-images.js
/**
 * Image Processing Script for Hyperinflation Banknote Visualizer
 *
 * This script processes original banknote images to create:
 * - Web-optimized versions (1200px width, ~200KB)
 * - Thumbnails (300px width, ~50KB)
 *
 * Requirements:
 *   npm install sharp
 *
 * Usage:
 *   node scripts/process-images.js
 */

const fs = require('fs');
const path = require('path');

// Check if sharp is installed
let sharp;
try {
    sharp = require('sharp');
} catch (e) {
    console.error('Sharp is not installed. Please run: npm install sharp');
    console.log('\nAlternatively, you can use the Python script: python scripts/process-images.py');
    process.exit(1);
}

// Configuration
const CONFIG = {
    inputDir: path.join(__dirname, '..', 'assets', 'bills', 'originals'),
    webDir: path.join(__dirname, '..', 'assets', 'bills', 'web'),
    thumbDir: path.join(__dirname, '..', 'assets', 'bills', 'thumbnails'),
    webWidth: 1200,      // Width for web display
    thumbWidth: 300,     // Width for thumbnails
    webQuality: 85,      // JPEG quality for web (0-100)
    thumbQuality: 80,    // JPEG quality for thumbnails
    format: 'jpeg',      // Output format (jpeg for smaller size)
    supportedFormats: ['.png', '.jpg', '.jpeg', '.webp', '.tiff']
};

// Ensure output directories exist
function ensureDirectories() {
    [CONFIG.webDir, CONFIG.thumbDir].forEach(dir => {
        if (!fs.existsSync(dir)) {
            fs.mkdirSync(dir, { recursive: true });
            console.log(`Created directory: ${dir}`);
        }
    });
}

// Get all image files from input directory
function getImageFiles() {
    if (!fs.existsSync(CONFIG.inputDir)) {
        console.error(`Input directory does not exist: ${CONFIG.inputDir}`);
        console.log('\nPlease extract the banknote images to: assets/bills/originals/');
        return [];
    }

    return fs.readdirSync(CONFIG.inputDir)
        .filter(file => {
            const ext = path.extname(file).toLowerCase();
            return CONFIG.supportedFormats.includes(ext);
        });
}

// Process a single image
async function processImage(filename) {
    const inputPath = path.join(CONFIG.inputDir, filename);
    const baseName = path.parse(filename).name;

    // Output paths (always JPEG for smaller files)
    const webPath = path.join(CONFIG.webDir, `${baseName}.jpg`);
    const thumbPath = path.join(CONFIG.thumbDir, `${baseName}.jpg`);

    try {
        const image = sharp(inputPath);
        const metadata = await image.metadata();

        console.log(`Processing: ${filename} (${metadata.width}x${metadata.height})`);

        // Create web version
        await sharp(inputPath)
            .resize(CONFIG.webWidth, null, {
                withoutEnlargement: true,
                fit: 'inside'
            })
            .jpeg({ quality: CONFIG.webQuality, progressive: true })
            .toFile(webPath);

        // Create thumbnail
        await sharp(inputPath)
            .resize(CONFIG.thumbWidth, null, {
                withoutEnlargement: true,
                fit: 'inside'
            })
            .jpeg({ quality: CONFIG.thumbQuality })
            .toFile(thumbPath);

        // Get output file sizes
        const webStats = fs.statSync(webPath);
        const thumbStats = fs.statSync(thumbPath);
        const originalStats = fs.statSync(inputPath);

        console.log(`  ✓ Web: ${(webStats.size / 1024).toFixed(0)}KB`);
        console.log(`  ✓ Thumb: ${(thumbStats.size / 1024).toFixed(0)}KB`);
        console.log(`  ✓ Original: ${(originalStats.size / 1024 / 1024).toFixed(1)}MB`);

        return {
            filename,
            success: true,
            originalSize: originalStats.size,
            webSize: webStats.size,
            thumbSize: thumbStats.size
        };

    } catch (error) {
        console.error(`  ✗ Error processing ${filename}: ${error.message}`);
        return {
            filename,
            success: false,
            error: error.message
        };
    }
}

// Main function
async function main() {
    console.log('╔════════════════════════════════════════════════════════╗');
    console.log('║  Hyperinflation Banknote Image Processor               ║');
    console.log('╚════════════════════════════════════════════════════════╝\n');

    ensureDirectories();

    const images = getImageFiles();

    if (images.length === 0) {
        console.log('No images found to process.');
        return;
    }

    console.log(`Found ${images.length} images to process\n`);
    console.log('─'.repeat(50));

    const results = [];
    for (const image of images) {
        const result = await processImage(image);
        results.push(result);
        console.log('');
    }

    // Summary
    console.log('─'.repeat(50));
    console.log('\n📊 Summary:');

    const successful = results.filter(r => r.success);
    const failed = results.filter(r => !r.success);

    console.log(`  ✓ Processed: ${successful.length}/${images.length}`);

    if (successful.length > 0) {
        const totalOriginal = successful.reduce((sum, r) => sum + r.originalSize, 0);
        const totalWeb = successful.reduce((sum, r) => sum + r.webSize, 0);
        const totalThumb = successful.reduce((sum, r) => sum + r.thumbSize, 0);

        console.log(`\n  Original total: ${(totalOriginal / 1024 / 1024).toFixed(1)}MB`);
        console.log(`  Web total: ${(totalWeb / 1024 / 1024).toFixed(1)}MB (${((1 - totalWeb/totalOriginal) * 100).toFixed(0)}% reduction)`);
        console.log(`  Thumbnails total: ${(totalThumb / 1024 / 1024).toFixed(1)}MB`);
    }

    if (failed.length > 0) {
        console.log(`\n  ✗ Failed: ${failed.length}`);
        failed.forEach(r => console.log(`    - ${r.filename}: ${r.error}`));
    }

    console.log('\n✅ Done!');
}

main().catch(console.error);

File: scripts/process-images.py
"""
Image Processing Script for Hyperinflation Banknote Visualizer

This script processes original banknote images to create:
- Web-optimized versions (1200px width, ~200KB)
- Thumbnails (300px width, ~50KB)

Requirements:
    pip install Pillow

Usage:
    python scripts/process-images.py
"""

import os
import sys
import io
from pathlib import Path

# Fix encoding for Windows
sys.stdout = io.TextIOWrapper(sys.stdout.buffer, encoding='utf-8', errors='replace')
sys.stderr = io.TextIOWrapper(sys.stderr.buffer, encoding='utf-8', errors='replace')

try:
    from PIL import Image
except ImportError:
    print("Pillow is not installed. Please run: pip install Pillow")
    sys.exit(1)

# Configuration
CONFIG = {
    'input_dir': Path(__file__).parent.parent / 'assets' / 'bills' / 'originals',
    'web_dir': Path(__file__).parent.parent / 'assets' / 'bills' / 'web',
    'thumb_dir': Path(__file__).parent.parent / 'assets' / 'bills' / 'thumbnails',
    'web_width': 1200,
    'thumb_width': 300,
    'web_quality': 85,
    'thumb_quality': 80,
    'supported_formats': {'.png', '.jpg', '.jpeg', '.webp', '.tiff', '.bmp'}
}


def ensure_directories():
    """Create output directories if they don't exist."""
    for dir_path in [CONFIG['web_dir'], CONFIG['thumb_dir']]:
        dir_path.mkdir(parents=True, exist_ok=True)
        print(f"Ensured directory: {dir_path}")


def get_image_files():
    """Get all image files from input directory."""
    input_dir = CONFIG['input_dir']

    if not input_dir.exists():
        print(f"Input directory does not exist: {input_dir}")
        print("\nPlease extract the banknote images to: assets/bills/originals/")
        return []

    return [
        f for f in input_dir.iterdir()
        if f.is_file() and f.suffix.lower() in CONFIG['supported_formats']
    ]


def resize_image(img, max_width):
    """Resize image maintaining aspect ratio."""
    if img.width <= max_width:
        return img.copy()

    ratio = max_width / img.width
    new_height = int(img.height * ratio)
    return img.resize((max_width, new_height), Image.Resampling.LANCZOS)


def process_image(image_path):
    """Process a single image, creating web and thumbnail versions."""
    base_name = image_path.stem
    web_path = CONFIG['web_dir'] / f"{base_name}.jpg"
    thumb_path = CONFIG['thumb_dir'] / f"{base_name}.jpg"

    try:
        with Image.open(image_path) as img:
            # Convert to RGB if necessary (for PNG with transparency)
            if img.mode in ('RGBA', 'P'):
                # Create white background
                background = Image.new('RGB', img.size, (255, 255, 255))
                if img.mode == 'P':
                    img = img.convert('RGBA')
                background.paste(img, mask=img.split()[-1] if img.mode == 'RGBA' else None)
                img = background
            elif img.mode != 'RGB':
                img = img.convert('RGB')

            original_size = image_path.stat().st_size
            print(f"Processing: {image_path.name} ({img.width}x{img.height})")

            # Create web version
            web_img = resize_image(img, CONFIG['web_width'])
            web_img.save(web_path, 'JPEG', quality=CONFIG['web_quality'], progressive=True)

            # Create thumbnail
            thumb_img = resize_image(img, CONFIG['thumb_width'])
            thumb_img.save(thumb_path, 'JPEG', quality=CONFIG['thumb_quality'])

            # Get output sizes
            web_size = web_path.stat().st_size
            thumb_size = thumb_path.stat().st_size

            print(f"  ✓ Web: {web_size // 1024}KB")
            print(f"  ✓ Thumb: {thumb_size // 1024}KB")
            print(f"  ✓ Original: {original_size / 1024 / 1024:.1f}MB")

            return {
                'filename': image_path.name,
                'success': True,
                'original_size': original_size,
                'web_size': web_size,
                'thumb_size': thumb_size
            }

    except Exception as e:
        print(f"  ✗ Error processing {image_path.name}: {e}")
        return {
            'filename': image_path.name,
            'success': False,
            'error': str(e)
        }


def main():
    print("╔════════════════════════════════════════════════════════╗")
    print("║  Hyperinflation Banknote Image Processor (Python)      ║")
    print("╚════════════════════════════════════════════════════════╝\n")

    ensure_directories()

    images = get_image_files()

    if not images:
        print("No images found to process.")
        return

    print(f"Found {len(images)} images to process\n")
    print("─" * 50)

    results = []
    for image_path in sorted(images):
        result = process_image(image_path)
        results.append(result)
        print()

    # Summary
    print("─" * 50)
    print("\n📊 Summary:")

    successful = [r for r in results if r['success']]
    failed = [r for r in results if not r['success']]

    print(f"  ✓ Processed: {len(successful)}/{len(images)}")

    if successful:
        total_original = sum(r['original_size'] for r in successful)
        total_web = sum(r['web_size'] for r in successful)
        total_thumb = sum(r['thumb_size'] for r in successful)

        print(f"\n  Original total: {total_original / 1024 / 1024:.1f}MB")
        print(f"  Web total: {total_web / 1024 / 1024:.1f}MB ({(1 - total_web/total_original) * 100:.0f}% reduction)")
        print(f"  Thumbnails total: {total_thumb / 1024 / 1024:.1f}MB")

    if failed:
        print(f"\n  ✗ Failed: {len(failed)}")
        for r in failed:
            print(f"    - {r['filename']}: {r['error']}")

    print("\n✅ Done!")


if __name__ == "__main__":
    main()
